spectator-web-spring/src/main/java/com/netflix/spectator/controllers/MetricsController.java

```diff
@@ -135,77 +128,81 @@ public ApplicationRegistry getMetrics(@RequestParam Map<String, String> filters)
-           } else {
-              have.addMeasurements(kind, entry.getValue());
-           }
-        }
-      }
-    }
-    return metricMap;
-  }
+      String kind = knownMeterKinds.computeIfAbsent(
+                         meter.id(), k -> meterToKind(sourceRegistry, meter));
 
-  /**
-   * Collect all the meter values matching the tag pattern.
-   */
-  public static void collectValues(
-      Map<Id, List<Measurement>> collection,
-      Meter meter,
-      MeasurementFilter filter) {
-    for (Measurement measurement : meter.measure()) {
-      if (!filter.keep(meter, measurement)) {
+      for (Measurement measurement : meter.measure()) {
+        if (!filter.test(measurement)) {
           continue;
+        }
+
+        String measurementName = measurement.id().name();
+        MetricValues have = metricMap.get(measurementName);
+        if (have == null) {
+          metricMap.put(measurementName, new MetricValues(kind, measurement));
+        } else {
+          have.addMeasurement(measurement);
+        }
       }
-      Id id = measurement.id();
-      collection.computeIfAbsent(id, k -> new ArrayList<Measurement>())
-          .add(measurement);
     }
+    return metricMap;
   }
 
   /**
    * Determine the type of a meter for reporting purposes.
+   *
+   * The accuracy of this method is sensiive to the registry value.
+   * In practice meters are often AggrMeter or CompositeMeter so it
+   * is not worth testing the class against the interface because it
+   * is rarely an actual kind. For lack of a direct way to query
+   * we will iterate over all the instances in the registry of a given
+   * type using the registry stream getters.
+   *
+   * 20160918(ewiseblatt):
+   * However, this doesnt work either unless the registry instance
+   * we have is exactly right and spring seems to get in the way.
+   * If the registry was a bean, you might need to autowire it.
+   * But it isnt clear if all the metrics are created through that
+   * registry. The Spectator.globalRegistry, for example, iterates over
+   * all the meters using the generic iterator, but the kind-specific
+   * stream functions are empty if the registry was created as a spring
+   * bean so you need to passing in an @autowired registry.
+   *
+   * @param registry
+   *    Used to provide suplemental information (e.g. to search for the meter).
+   *
+   * @param meter
+   *    The meters whose kind we want to know.
+   *
+   * @return
+   *    A string such as "Counter". If the type cannot be identified as one of
+   *    the standard Spectator api interface variants, then the simple class name
+   *    is returned.
    */
-  public static String meterToKind(Meter meter) {
+  public static String meterToKind(Registry registry, Meter meter) {
     String kind;
-    if (meter instanceof Counter) {
+    if (registry.counters()
+              .anyMatch(m -> m.id().equals(meter.id()))) {
```

Comment:
I'm hoping you'll add support into the interface so I can tell what type of meter I have directly. I'm only doing these iterations because it is the last resort (in practice almost everything is a CompositeMeter class so I took out checking instances).

I can cache the results in a Map<Id, String>.

I do that (Boolean) in StackdriverWriter:
  private Map<Id, Boolean> idToTimer = new HashMap<Id, Boolean>();
  public boolean meterIsTimer(Registry registry, Meter meter) {
    return idToTimer.computeIfAbsent(meter.id(), k -> {
        return registry.timers().anyMatch(m -> m.id().equals(meter.id()));
    });
  }